### PR TITLE
Александрычев А.С. 381806-2

### PR DESCRIPTION
**Pull request by Alex-Andr-19**

### CODE REVIEW

File: main.cpp [hunk 1]
```diff
@@ -18,13 +18,29 @@ using namespace clang::ast_matchers;
...
+
 public:
-    CastCallBack(Rewriter& rewriter) {
-        // Your code goes here
-    };
+    CastCallBack(Rewriter& _rewriter) : rewriter_(_rewriter) {};
 
     void run(const MatchFinder::MatchResult &Result) override {
-        // Your code goes here
+        const auto *res = Result.Nodes.getNodeAs<CStyleCastExpr>("cast");
```
Comment (eshulankina) [inline main.cpp:28]: What if `res` is a null pointer? How to protect the program from crashing in such cases?

File: main.cpp [hunk 2]
```diff
@@ -18,13 +18,29 @@ using namespace clang::ast_matchers;
...
+
+        auto ReplRange = CharSourceRange::getCharRange(res->getLParenLoc(), res->getSubExprAsWritten()->getBeginLoc());
+
+        StringRef DestTypeString = Lexer::getSourceText(CharSourceRange::
+        getTokenRange(res->getLParenLoc().getLocWithOffset(1), res->getRParenLoc().getLocWithOffset(-1)), *Result.SourceManager, Result.Context->getLangOpts());
+
+        std::string TypeText = ("static_cast<" + DestTypeString + ">").str();
+        const auto* Ignore = res->getSubExprAsWritten()->IgnoreImpCasts();
+
+        if (!isa<ParenExpr>(Ignore)) {
```
Comment (eshulankina) [inline main.cpp:39]: Please describe what this check does, why it was added here.

File: main.cpp [hunk 3]
```diff
@@ -18,13 +18,29 @@ using namespace clang::ast_matchers;
...
+        const auto *res = Result.Nodes.getNodeAs<CStyleCastExpr>("cast");
+        SourceManager &SourceM = *Result.SourceManager;
+
+        auto ReplRange = CharSourceRange::getCharRange(res->getLParenLoc(), res->getSubExprAsWritten()->getBeginLoc());
+
+        StringRef DestTypeString = Lexer::getSourceText(CharSourceRange::
+        getTokenRange(res->getLParenLoc().getLocWithOffset(1), res->getRParenLoc().getLocWithOffset(-1)), *Result.SourceManager, Result.Context->getLangOpts());
+
+        std::string TypeText = ("static_cast<" + DestTypeString + ">").str();
+        const auto* Ignore = res->getSubExprAsWritten()->IgnoreImpCasts();
```
Comment (eshulankina) [inline main.cpp:37]: What is the difference between `getSubExprAsWritten()` and `getSubExpr()`? Please provide an example.
What is the purpose of using the `IgnoreImpCasts()` method?

File: main.cpp (hunk 1 above)
Comment (Alex-Andr-19) [reply main.cpp:28]: Function `.getNodeAs` can return **NULL** if there was no node bound to ID or if there is a node but it cannot be converted to the specified type. To work around this problem, we can catch it before definition `res` and catch that.
`if (CastExpr == nullptr) return;`

File: main.cpp (hunk 3 above)
Comment (Alex-Andr-19) [reply main.cpp:37]: `getSubExpAsWritten()` : Retrieve the cast subexpression as it was written in the source code, looking through any implicit casts or other intermediate nodes introduced by semantic analysis.
`getSubExp()` is just reference by `getSubExpAsWritten()`.
So using the `getSubExpAsWritten()` will be safer than `getSubExp()`.

I used `IgnoreImpCasts()` to skip past any implicit casts which might surround this expression until reaching a fixed point.

File: main.cpp (hunk 2 above)
Comment (Alex-Andr-19) [reply main.cpp:39]: This check using when brackets are missed in expression. It is nesseccary just when brackets are used in call of function.
For exemple:
    `int i = static_cast<int>f;`
It have to be:
    `int i = static_cast<int>(f);`

File: main.cpp (hunk 3 above)
Comment (eshulankina) [reply main.cpp:37]: If you try to print the AST of the `test.cpp` example, you will see that you have several `ImplicitCastExpr` nodes:
image

Depending on which approaches you use to get the subexpression node, you will get different result nodes for the `test.cpp` example:

Code:
```cpp
// getSubExpr() retrieves the first cast subexpression
llvm::outs() << "getSubExpr(): " << castNode->getSubExpr()->getStmtClassName();

// getSubExprAsWritten() retrieves the cast subexpression as it was written in the source
// code, looking through any implicit casts or other intermediate nodes
// introduced by semantic analysis.
llvm::outs() << "getSubExprAsWritten(): " << castNode->getSubExprAsWritten()->getStmtClassName();

// getSubExpr() retrieves the first cast subexpression.
// ignoreImpCasts() skips past any implicit casts which might surround this expression until
// reaching a fixed point. It skips the following AST nodes: ImplicitCastExpr, FullExpr.
llvm::outs() << "getSubExpr()->ignoreImpCasts(): " << castNode->getSubExpr()->IgnoreImpCasts()->getStmtClassName();
```
Output:
```bash
 getSubExpr(): ImplicitCastExpr 
 getSubExprAsWritten(): DeclRefExpr  
 getSubExpr()->ignoreImpCasts(): DeclRefExpr
```

File: main.cpp (hunk 1 above)
Comment (eshulankina) [reply main.cpp:28]: There is no `CastExpr` in your code.

File: main.cpp [hunk 4]
```diff
@@ -18,13 +18,29 @@ using namespace clang::ast_matchers;
...
+
+        if (!isa<ParenExpr>(Ignore)) {
+            TypeText.push_back('(');
+            rewriter_.InsertText(Lexer::getLocForEndOfToken(Ignore->getEndLoc(), 0, *Result.SourceManager, Result.Context->getLangOpts()), ")");
```
Comment (eshulankina) [inline main.cpp:41]: Please describe the reason for using `getLocForEndOfToken()` here.  
What would be the difference if `Ignore->getEndLoc().getLocOffset(1)` was used here.

File: main.cpp (hunk 1 above)
Comment (Alex-Andr-19) [reply main.cpp:28]: By this way I would like to express `res` like `CStyleCastExpr` type

Comment (Alex-Andr-19) [reply main.cpp:28]: Of course condition should look like:
`if (res == nullptr) return;`

File: main.cpp (hunk 4 above)
Comment (eshulankina) [reply main.cpp:41]: No answers still.

**Pull request closed without merging**